### PR TITLE
feat: ils auto-tuning for departure and approach

### LINKED ISSUE #3696: [BUG] Plane Not Filling in ILS frequency in RAD/NAV page When spawned in on ILS approach
**Mod Version**
{
  "built": "2021-02-23T05:59:34+00:00",
  "ref": "fbw",
  "sha": "fa3e284a9e4a0aa07c0582ca5e2a88a9689b1f9f",
  "actor": "aguther",
  "event_name": "manual"
}

**Describe the bug**
When spawning in on approach, the plane does not want to follow the ILS and instead follows NAV

**To Reproduce**
1. Spawn in on approach at any airport with the ILS approach (in the main menu)
2. Turn on AP, and Approach mode, check PFD and see if it catches the LOC and GS

**Expected behavior**

The plane should fly the ILS when spawned in on the ILS

**Actual behavior**
Spawn in on approach and let the AP fly the ILS without having to reprogram the MCDU.

**References**
A video of me spawning in at CYYZ and the ILS approach to 33R

https://www.youtube.com/watch?v=PyOS1WpUFJY

Seems as though the ILS frequency isn't inputed: 

image

**Additional context**

Was this working before/when did the issue start occurring?
The issue started happening 2 weeks ago

Is this a problem in the vanilla unmodded game?

No

Discord username: Darlack#3281

### LINKED ISSUE #4729: Wrong approach tuned.
### Aircraft Version

Development

### Build info

```json
{
  "built": "2021-05-06T07:02:50+00:00",
  "ref": "master",
  "sha": "73ed584d7be313f72b2869618582d8af77fca1e0",
  "actor": "Benjozork",
  "event_name": "manual"
}
```

### Describe the bug

Loaded ILS01Z approach at BIKF. Route loaded fine on the ND, however, the ILS tuned to the freq for ILS RWY10...

image

### Expected behavior

The ILS should have tuned for RWY01.

### Steps to reproduce

Fly to BIKF, example route:

EGPH/24 GOSAM1C GOSAM DCT GOW DCT BRUCE DCT GOMUP DCT BREKI DCT BASLU BASLU3N BIKF/01

image

Select the ILS01Z approach in the FMGS.

Fly the approach.

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

solarisfire#0001

### PR DESCRIPTION
**Pull request by tracernz**
Fixes #3696
Fixes #4729
Half of #4962

## Summary of Changes

ILS is autotuned on departure, and now realiably on approach (reliably). Proper behaviour is to autotune the departure runway ILS during preflight and takeoff phases, and then approach ILS after that, if and when runways are selected in the F-PLN.

## Screenshots (if necessary)

Departure (not the ILS is way off centre at Auckland... either the scenery (add-on) or the navdata is wrong):

## Additional context

I will fully replace this code at a later date with a complete rewrite of nav radio tuning logic (part of new typescript FMGC)... but this is a good intermediate step.

Discord username (if different from GitHub):

## Testing instructions

Need to do at least one full flight to check correct tuning at both ends. Also try spawning on approach. Try changing runway after ILS is already tuned. Try some manual tuning, and then clearing to see that auto-tune takes over.

This needs at least a test with and without Navigraph nav data. I have only tried with Navigraph data installed.

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Comment by frankkopp** [comment]
I know it is not ready to test yet, but I love this one. Works fine for the simwings Munich airport now where ILS was not automatically set. Also like the auto setting at departure and after. Will do a full report once it is ready to test.

**Comment by tracernz** [comment]
I have done a quick test with MSFS default navdata, and it works there too. 😁

File: flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/A320_Neo_CDU_NavRadioPage.js [hunk 1]
```diff
@@ -114,20 +114,14 @@ class CDUNavRadioPage {
...
                 if (mcdu._ilsFrequencyPilotEntered) {
-                    ilsFrequencyCell = ilsIdentStr + "/" + mcdu.ilsFrequency.toFixed(2);
-                    ilsCourseCell = "{inop}____{end}";
-                } else {
-                    ilsFrequencyCell = ilsIdentStr + "{small}" + "/" + mcdu.ilsFrequency.toFixed(2) + "{end}";
-                    ilsCourseCell = "{small}" + ilsIdent.course.toFixed(0).padStart(3, "0") + "{end}";
+                    const ilsIdent = mcdu.radioNav.getILSBeacon(1);
+                    ilsFrequencyCell = `{small}${ilsIdent.ident.trim().padStart(4, "\xa0")}{end}/${mcdu.ilsFrequency.toFixed(2)}`;
+                    ilsCourseCell = "{small}F" + ilsIdent.course.toFixed(0).padStart(3, "0") + "{end}";
+                } else if (mcdu.ilsAutoTuned) {
```
Comment (beheh) [inline flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/A320_Neo_CDU_NavRadioPage.js:122]: Is there a scenario where `mcdu.ilsFrequency` is non-zero, yet both `_ilsFrequencyPilotEntered` and `ilsAutoTuned` are false-ish? Aren't they somehow exclusive?

Comment (tracernz) [reply flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/CDU/A320_Neo_CDU_NavRadioPage.js:122]: I don't think there is, but I do prefer an explicit condition, to protect against future changes (e.g. somebody adds a bit of code to sync mcdu.ilsFrequency with the sim on spawn)... that said, this is all ripped out and replaced in my upcoming nav radio branch so it won't be around long anyway... Can I leave it as-is for now?

**Comment by Illuminati270** [comment]
Hi guys, this one says it's missing a review, but there is no label "Ready to test". Is it incorporated to the most recent build? Or should I download it parallel to it?

**Comment by scoopscode** [comment]
> Hi guys, this one says it's missing a review, but there is no label "Ready to test". Is it incorporated to the most recent build? Or should I download it parallel to it?

A review is made by a maintainer of FBW. Although there is a QA Tier 2 label, that does not mean the PR is ready for testing.

**Comment by frankkopp** [comment]
> Hi guys, this one says it's missing a review, but there is no label "Ready to test". Is it incorporated to the most recent build? Or should I download it parallel to it?

Review != Testing :) - it means code review

I'm also waiting for this on to become RTT as my home base Munich really could use this feature.

**Review by MisterChocker — APPROVED** [review]
Code wise looking god.

@beheh are you also satisfied?

**Comment by frankkopp** [comment]
Quality Assurance Tester Report

Discord: Cdr_Maverick#6475
Object of testing: #4973
Tier of Testing: 2
Date: 11.6.2021

Testing Process:
> Need to do at least one full flight to check correct tuning at both ends. 

Full flight EDDS-EDDM (EDDM simwings ==> did not autopopulate ILS freq before this PR, Navigraph)

> Try changing runway after ILS is already tuned. 

Start with 26L and change during flight to 26R

> Try some manual tuning, and then clearing to see that auto-tune takes over.

==> OK ==> see videos

> Also try spawning on approach. 

Spawned on EDDM 26R approach with ILS26R approach configured in MSFS planning window

==> OK (no video)
(also did one without ILS in MSFS planning window -> no ILS tuned but could be tuned manually or by selecting a RWY)

> This needs at least a test with and without Navigraph nav data. I have only tried with Navigraph data installed.

Uninstalled Navigraph data and did part of the test again.

Start on runway EDDM 26R dest EDDF (as EDDS is not available without Navigraph)
ILS25L ==> OK
ILS25 ==> NOT - but I think this is a MSFS data issue (this is ILS25C but the C is not shown)

==> OK

Negatives: None
Testing Results: Passed
Conclusions: LGTM

Media: 

F-Plan Setup: https://youtu.be/u1I3G97eg-Y
TakeOff: https://youtu.be/R0eCnDF10s4
Change Runway in air: https://youtu.be/-aA50nU2Itw

**Review by beheh — APPROVED** [review]

**Comment by marcman86** [comment]
Quality Assurance Tester Report
Discord:marcman86#4907
Object of testing: #4973
Tier of Testing: 1
Date: 11/06/2021

Testing Process 1: 
- Spawned on runway 25R at EDDB
- In MCDU selected 25R as departure
- Checked RAD/NAV > 109.9
- Took off
- Arrival 07L at EDDB
- Checked RAD/NAV > 110.7
- Autoland

Testing Process 2: 
- Spawned in air on approach on EDDB07L
- Checked RAD/NAV > 110.7
- Changed approach to 25R
- Checked RAD/NAV > 109.9

Negatives: -

Testing Results: Passed
Conclusions: Works

Hope this helps :-)

>

**Comment by TheSquirtingPantin** [comment]
Quality Assurance Tester

Discord: Remi#8871
Object of testing: #4973
Tier of Testing: 2
Date: 11/06/2021

Testing Process:
1) CD at KMIA
2) Full flight from KDEN to KMIA.
3) Runway changed on ground from 34L to 08

Negatives: NONE
Testing Results: Passed. 

Conclusions: LGTM. Autotune correct during the departure phase. Good adjustment during runway change. No RWY mode displayed on FMA or active during takeoff roll. Multiple approaches loaded during approach, RAD/NAV autotune logic correct
Media: N/A

**Pull request merged**